framework/sandstone.cpp

```diff
@@ -2611,70 +2610,93 @@ static int exec_mode_run(int argc, char **argv)
-
-    if (topo.packages.size() == 1) {
-        result.push_back(it->id);
-        return result;
-    }
-
     // backup the original verbosity
     orig_verbosity = sApp->verbosity;
     sApp->verbosity = -1;
 
-    for (; it != topo.packages.end(); disabled_sockets.push_back(it->id), ++it) {
-        vector<Socket>::iterator eit = it;
+    // first "non-empty" package found in the topology
+    vector<Socket>::iterator first_it = topo.packages.end();
+    // first "non-empty" package being "removed" from the list
+    vector<Socket>::iterator removed_it = topo.packages.end();
+    // last "non-empty" package being added
+    vector<Socket>::iterator last_it = topo.packages.end();
+
+    vector<Socket>::iterator it = topo.packages.begin();
+    while (it != topo.packages.end()) {
+        vector<Socket>::iterator eit;
         LogicalProcessorSet run_cpus = {};
-        int k = 0;
 
-        for (; eit != topo.packages.end(); ++eit)
-            run_cpus.add_package(*eit);
+        // merge all CPUs from all packages in the list (w/ respect to the removed packages)
+        for (eit = it; eit != topo.packages.end(); ++eit) {
+            if (!(*eit).cores.empty()) {
+                if (first_it == topo.packages.end()) {
+                    first_it = eit;
+                }
+                last_it = eit;
+                run_cpus.add_package(*eit);
+            }
+        }
 
-        sApp->enabled_cpus = run_cpus;
-        do {
-            ret = run_tests_on_cpu_set(triage_tests, run_cpus);
-        } while (!ret && ++k < sApp->retest_count);
+        if (!run_cpus.empty()) {
+            sApp->enabled_cpus = run_cpus;
+            // should we run any tests if retest_count == 0??
+            // Keep the loop running only for SUCCESSes
+            int k = 0;
+            do {
+                ret = run_tests_on_cpu_set(triage_tests, run_cpus);
+            } while ((ret == EXIT_SUCCESS) && ++k < sApp->retest_count);
 
-        if (ret) ever_failed = true; /* we've seen a failure */
+            if (ret > EXIT_SUCCESS) ever_failed = true; /* we've seen a **FAILURE** */
 
-        if (!ret && ever_failed) {
-            // the last socket removed is the main suspect
-            result.push_back(disabled_sockets.at(disabled_sockets.size() - 1));
-            break;
+            if ((ret == EXIT_SUCCESS) && ever_failed) {
+                // the last socket removed is the main suspect
+                result.push_back((*removed_it).id);
+                break;
+            }
         }
-    }
 
-    if (ret) { // failed on the last socket as well, so it's the main suspect
-        // re-run on the first to make sure the last one is faulty
-        LogicalProcessorSet run_cpus = {};
-        int k = 0;
+        // find first non-empty package to "remove"
+        while (it != topo.packages.end()) {
+            if (!(*it).cores.empty()) {
+                removed_it = it;
+                it++;
```

Comment:
Please write `++it` for iterators. The two functions are slightly different:
```c++
      _GLIBCXX20_CONSTEXPR
      __normal_iterator&
      operator++() _GLIBCXX_NOEXCEPT
      {
	++_M_current;
	return *this;
      }

      _GLIBCXX20_CONSTEXPR
      __normal_iterator
      operator++(int) _GLIBCXX_NOEXCEPT
      { return __normal_iterator(_M_current++); }
```

Though the compiler should optimise the difference out of existence.

Also, both sides of the `if` advance the iterator. Might it not be simpler to write a for loop?
```c++
for ( ; it != topo.packages.end(); ++it) {
    if (!it->cores.empty()) {
        removed_it = it;
        break;
}
```
or if you like functional programming:
```c++
removed_it = std::find_if(it, topo.packages.end(), [](auto eit) { return !eit->cores.empty(); });
```

Reply:
C++ construct is best here, thanks.